src/primitives/deferred.ts

```diff
@@ -0,0 +1,70 @@
+/*!
+ * This source file is part of the EdgeDB open source project.
+ *
+ * Copyright 2020-present MagicStack Inc. and the EdgeDB authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export class Deferred<T> {
+  private _promise: Promise<T | undefined>;
+  private _resolve?: (value?: T | PromiseLike<T> | undefined) => void;
+  private _reject?: (reason?: any) => void;
+  private _result: T | PromiseLike<T> | undefined;
+  private _done: boolean;
+
+  get promise(): Promise<T | undefined> {
+    return this._promise;
+  }
+
+  get done(): boolean {
+    return this._done;
+  }
+
+  get result(): T | PromiseLike<T> | undefined {
+    if (!this._done) {
+      throw new Error("The deferred is not resolved.");
+    }
+    return this._result;
+  }
+
+  async setResult(value?: T | PromiseLike<T> | undefined): Promise<void> {
+    while (!this._resolve) {
+      await new Promise<void>((resolve) => process.nextTick(resolve));
+    }
+    this._resolve(value);
+  }
+
+  async setFailed(reason?: any): Promise<void> {
+    while (!this._reject) {
+      await new Promise<void>((resolve) => process.nextTick(resolve));
+    }
+    this._reject(reason);
+  }
+
+  constructor() {
+    this._done = false;
+    this._reject = undefined;
+    this._resolve = undefined;
+
+    this._promise = new Promise((resolve, reject) => {
+      this._reject = reject;
+
+      this._resolve = (value?: T | PromiseLike<T> | undefined) => {
+        this._done = true;
+        this._result = value;
+        resolve(value);
+      };
+    });
+  }
+}
```

Comment:
Let's replace this entire file with this:


```typescript
/*!
 * This source file is part of the EdgeDB open source project.
 *
 * Copyright 2021-present MagicStack Inc. and the EdgeDB authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

type Resolver<T> = (value: T | PromiseLike<T>) => void;
type Rejector = (reason: any) => void;

export default class Event {
  private _promise: Promise<boolean>;
  private _resolve: Resolver<boolean>;
  private _reject: Rejector;
  private _done: boolean;

  async wait(): Promise<void> {
    await this._promise;
  }

  /** @internal */
  then(...args: any[]): any {
    // The mere presense of this method will trip TS if one awaits on
    // an Event object directly.
    throw new Error(
      "Event objects cannot be awaited on directly; use Event.wait()"
    );
  }

  get done(): boolean {
    return this._done;
  }

  set(): void {
    if (this._done) {
      throw new Error("emit(): the Event is already set");
    }
    this._resolve(true);
  }

  setError(reason: any): void {
    if (this._done) {
      throw new Error("emitError(): the Event is already set");
    }
    this._reject(reason);
  }

  constructor() {
    this._done = false;

    let futReject: Rejector | null = null;
    let futResolve: Resolver<boolean> | null = null;

    this._promise = new Promise<boolean>((resolve, reject) => {
      futReject = (reason) => {
        this._done = true;
        reject(reason);
      };

      futResolve = (value) => {
        this._done = true;
        resolve(value);
      };
    });

    if (!futReject || !futResolve) {
      // Impossible per the spec.
      throw new Error("Promise executor was not called synchronously");
    }

    this._reject = futReject;
    this._resolve = futResolve;
  }
}

export async function sleep(duration: number): Promise<void> {
  const sig = new Event();
  setTimeout(() => sig.set(), duration);
  await sig.wait();
}
```

This is from my rewrite attempt. The implementation is streamlined.

I also recommend fixing the current mess in raw connection with `this.messageWaiterResolve` and `this.this.messageWaiterResolve` by replacing them with single `this.messageWaiter = new Event;` (same for `this.connWaiter`)

Reply:
Done in https://github.com/edgedb/edgedb-js/commit/bfde9604d1c8cc2097fe3af85669abc6e85297ba + https://github.com/edgedb/edgedb-js/commit/3545a3cb7b6f8630f372e205ef37b44e498eee21